source/features/preserve-text-messages.js

```diff
@@ -14,85 +24,124 @@ function onDMDialogOpen() {
 			}
 		}
 
+		/** If Direct message modal is not open (hence also when it gets closed); */
 		if (!DMDialogOpen) {
-			const savedMessages = await browser.storage.local.get();
-			const idsOfMessagesToRemove = [];
+			const {messages: savedMessages} = await getFromLocalStorage('messages');
 
-			for (const id in savedMessages) {
-				if (isEmptyMsgInput(savedMessages[id])) {
-					idsOfMessagesToRemove.push(id);
-				}
+			if (!savedMessages) {
+				return;
 			}
 
-			await browser.storage.local.remove(idsOfMessagesToRemove);
+			// Select messages which are not empty
+			const idsOfMgsToSave = Object.keys(savedMessages)
+				.filter(id => !isEmptyMsgInput(savedMessages[id]));
+
+			// Empty messages are discarded before re write to local storage
+			const updatedMessages = {
+				messages: pick(savedMessages, idsOfMgsToSave)
+			};
+
+			setToLocalStorage(updatedMessages);
 		}
 	}, {attributes: true});
 }
 
 function onDMOpen() {
-	const conversationOptions = {
+	const DMOpenMutationOptions = {
 		attributes: true,
 		attributeFilter: ['class']
 	};
 
-	observeEl('.DMConversation', mutations => {
+	observeEl(DMConversationSelector, mutations => {
 		let conversationId;
 		let messageContainer;
 
 		for (const mutation of mutations) {
 			if (mutation.target.classList.contains('DMActivity--open')) {
 				conversationId = getConversationId();
-				messageContainer = $('#tweet-box-dm-conversation');
+				messageContainer = $(DMTextBoxSelector);
 				if (conversationId) {
-					onMessageChange(messageContainer);
+          fetchStoredMessage(conversationId, messageContainer);
+          onMessageChange(messageContainer);
 				}
 			}
-			if (conversationId) {
-				fetchStoredMessage(conversationId, messageContainer);
+			break;
+		}
+	}, DMOpenMutationOptions);
+}
+
+/** When someone deletes the message; remove it from localstorage */
+export function onDMDelete() {
+	const DMDeleteMutationOptions = {
+		childList: true,
+		subtree: true,
+		attributes: true
+	};
+
+	observeEl('body', async mutations => {
+		const {messages: savedMessages} = await getFromLocalStorage('messages');
+
+		if (!savedMessages) {
+			return;
+		}
+
+		for (const mutation of mutations) {
+			if (mutation.target.id === 'confirm_dialog') {
+				const conversationId = getConversationId();
+				$('#confirm_dialog_submit_button').on('click', () => {
+					// Get all the ids of messages that are not being deleted
+					const idsOfMsgToReSave = Object.keys(savedMessages)
+						.filter(id => id !== conversationId);
+
+					const updatedMessages = {
+						messages: pick(savedMessages, idsOfMsgToReSave)
```

Comment:
`Object.assign` should work here and we would'nt need `lodash.pick` as a dependency. Have you tried it?